webui: Use ProgressStepper on the installation progress screen

M4rtinK opened the pull request:
Use the PatternFly ProgressStepper component instead of a progress bar
on the installation progress screen.

**TODO:**
- [x] remove **Installation progress** in upper left
- [x] unit tests

![TestInstallationProgress-testBasic-installation-progress-step-fail-pixels](https://user-images.githubusercontent.com/14921356/162498357-4a388108-4ab2-486e-a4df-75ddb0a2f577.png)

M4rtinK commented:
Currently looks like this:
![webui_progress_stepper_3](https://user-images.githubusercontent.com/829558/162032758-8041437f-ccf0-437e-9306-2ac112693d4c.png)

KKoukiou commented on ui/webui/src/components/installation/InstallationProgress.jsx:
```diff
@@ -59,6 +58,25 @@ export class InstallationProgress extends React.Component {
                                             this.props.onAddErrorNotification
                                         );
                             }
+                            console.log("MESSAGE");
```
please remove console log

M4rtinK replied:
Removed. :)

KKoukiou commented on ui/webui/src/components/installation/InstallationProgress.jsx:
```diff
@@ -59,6 +58,25 @@ export class InstallationProgress extends React.Component {
+                                this.setState({ currentProgressStep: 0 });
+                            }
+                            // payload
+                            if (step === 4) {
```
You 'd rather use else if.

In this way if won't go over all conditions, if the first is truthy,

KKoukiou commented on ui/webui/src/components/installation/InstallationProgress.jsx:
```diff
@@ -79,8 +98,15 @@ export class InstallationProgress extends React.Component {
+
+        const progressSteps = [
+            { title: "Storage configuration", id: "installation-progress-step-storage" },
+            { title: "Payload installation", id: "installation-progress-step-paylod" },
```
s/paylod/payload

M4rtinK replied:
Fixed, thanks! :)

KKoukiou commented on ui/webui/src/components/installation/InstallationProgress.jsx:
```diff
@@ -97,31 +123,40 @@ export class InstallationProgress extends React.Component {
+                  <ProgressStepper isCenterAligned>
+                      {progressSteps.map((progressStep, index) => {
+                          let variant = "pending";
+                          const ariaLabel = "pending step";
```
aria labels should get translated normally. It's read by screen readers.

M4rtinK replied:
Thanks, I was wondering what all they are used for. BTW, I think we will need to add them in more places later on.

M4rtinK replied:
Anyway, fixed. :)

KKoukiou commented on ui/webui/src/components/installation/InstallationProgress.jsx:
```diff
@@ -97,31 +123,40 @@ export class InstallationProgress extends React.Component {
+                                key={index}
+                                variant={variant}
+                                isCurrent={index === currentProgressStep}
+                                aria-label={ariaLabel}
```
Can you sort the properties :) ? Just for the nice looks.

KKoukiou commented on ui/webui/src/components/installation/InstallationProgress.jsx:
```diff
@@ -97,31 +123,40 @@ export class InstallationProgress extends React.Component {
+            <EmptyStatePanel
+              icon={icon}
+              loading={!icon}
+              paragraph={
```
Since the paragraph content is SO long, you can store it in a variable and then use:

`paragraph={installationProgressStepper}`

M4rtinK replied:
Thanks for the suggestion but I think I'd like to keep the amount of indirection low in the rendering section, at least for now. :)

KKoukiou commented on ui/webui/src/components/installation/InstallationProgress.jsx:
```diff
@@ -69,6 +87,7 @@ export class InstallationProgress extends React.Component {
                         });
                         taskProxy.addEventListener("Succeeded", () => {
                             this.setState({ status: "success" });
```
You can do:
`this.setState({ status: "success", currentProgressStep: 4 });`

KKoukiou replied:
remember every time you call setState the component 're-renders'. React is smart enough to group setStates that are directly after each each, but if you just group them in the code as well, it's a good practice.

jstodola commented on ui/webui/src/components/installation/InstallationProgress.jsx:
```diff
@@ -59,6 +58,25 @@ export class InstallationProgress extends React.Component {
+                            console.log(step);
+                            console.log(message);
+                            // storage
+                            if (step <= 3) {
```
Will these magical constants be addressed somehow in the future? If yes, there should be a FIXME comment.

M4rtinK replied:
Oh, right, that's a really good point and really should have been noted in the commit message or a code comment. 

So basically the way how the progress stepping is implemented right now is rather tightly coupled to the MVP use case, primarily to get feedback for using the ProgressStepper in the Web UI. If the feedback is positive and we keep it, this will definitely have a proper API. 

Likely both for getting a list of phases (as those can very well change based on whats being installed) as well as proper signals when the installation process switches to the next phase.

jstodola commented on ui/webui/src/components/installation/InstallationProgress.jsx:
```diff
@@ -59,6 +58,25 @@ export class InstallationProgress extends React.Component {
+                            if (step >= 5 && step <= 11) {
+                                this.setState({ currentProgressStep: 2 });
+                            }
+                            // bootloader
```
Is the bootloader step really the last step? What about copying logs, for example?
I'm just wondering if bootloader configuration really deserves it's own step, IMO it's just one of the "configuration" substeps, like creating the users or generating initramfs.

M4rtinK replied:
Agreed, this is just implemented based on the mockup, which has bootloader as the last step. AFAIK we have not yet really discussed the step groupings in detail and the current progress steps should be thus considered preliminary. :)

VladimirSlavik replied:
Bootloader currently generates something and takes forever, so it makes some sense to have it as a major thing.

M4rtinK commented:
PR updated based on feedback.

KKoukiou reviewed (commented):
A final important thing. The mpckups have the step messages as well. https://marvelapp.com/prototype/i1cb8je/screen/85836244 


This previous implementation also did, so it would be a regression to not show them now. Can you please integrate them?

M4rtinK commented:
> A final important thing. The mpckups have the step messages as well. https://marvelapp.com/prototype/i1cb8je/screen/85836244
> 
> This previous implementation also did, so it would be a regression to not show them now. Can you please integrate them?
Oh right, that was actually not clear to me how this is supposed to work so I have not added it yet. Basically in the mockup there are just placeholder texts for all of the steps so it was not clear to me what actually goes there. 

If this is a static "This stuff does install foo into bar." or if it shows the current status message for the step in progress ? Then I think only the current step should show the message.

M4rtinK commented:
So I've dropped the text in upper left and added the current step text to below the currently active step:
![webui_progress_stepper_4](https://user-images.githubusercontent.com/829558/162227684-16f2d227-6fd8-4515-836a-7cf5686427cd.png)
Looking quite nice I would say. :)

M4rtinK commented:
The dynamic status message under progress step turns out to be problematic - changing string lenghts make the progress steps jump around. :P

So lets try just showing the status text below the progress stepper instead:

![webui_progress_stepper_5](https://user-images.githubusercontent.com/829558/162246444-b4b2ce57-53c7-44f3-9ae0-e8099609dec2.png)

KKoukiou commented on ui/webui/src/components/installation/InstallationProgress.jsx:
```diff
@@ -97,31 +121,45 @@ export class InstallationProgress extends React.Component {
+                              return (
+                                  <ProgressStep
+                                    aria-label={ariaLabel}
+                                    id={idPrefix + "-progress-step-" + index}
```
The idPrefix is installation-progress so the final here would be, installation-progress-progress-step-
You can remove the progress from the middle string.

M4rtinK replied:
Oh, right, fixed. :)

KKoukiou commented on ui/webui/src/components/installation/InstallationProgress.jsx:
```diff
@@ -97,31 +121,45 @@ export class InstallationProgress extends React.Component {
+                              );
+                          })}
+                      </ProgressStepper>
+                      <Text component={TextVariants.p}>
```
I am almost sure, that for the styles for the Text to be applied, you need to wrap it in a TextContent.

M4rtinK replied:
Lets do that. :)

KKoukiou commented on ui/webui/src/components/installation/InstallationProgress.jsx:
```diff
@@ -97,31 +121,45 @@ export class InstallationProgress extends React.Component {
         }
 
         return (
-            <Bullseye className={idPrefix + "-status-" + status}>
```
This ID was used by the custom CSS to make icons having colors https://github.com/rhinstaller/anaconda/blob/master/ui/webui/src/components/installation/InstallationProgress.scss#L8

Please add a <div id={this id } wrapper for the styles to continue working.

M4rtinK replied:
Interesting, should be fixed now. :)

M4rtinK replied:
Or at least in theory, icon is still grey for some reason. :P

KKoukiou commented on ui/webui/src/components/installation/InstallationProgress.jsx:
```diff
@@ -79,8 +96,15 @@ export class InstallationProgress extends React.Component {
+        const { steps, currentProgressStep, status, message } = this.state;
+
+        const progressSteps = [
+            { title: _("Storage configuration"), id: "installation-progress-step-storage" },
```
You can use the idPrefix here.

id: idPrefix + "-step-storage"

KKoukiou reviewed (approved):
If the tests also agree

VladimirSlavik reviewed (approved):
Some of the commits need squashing, otherwise LGTM.

KKoukiou commented:
/kickstart-test --testtype smoke

KKoukiou reviewed (commented):
Congratulations :) Great UI improvement!
